### PR TITLE
feat: implement and integrate counter-based CCC

### PR DESCRIPTION
**Pull request by omerfirmak**
## 1. Purpose or design rationale of this PR

this PR integrates counter-based CCC and reorging logic that would be triggered on circuit overflows

## 2. PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [x] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

## 3. Deployment tag versioning

Has the version in `params/version.go` been updated?

- [ ] This PR doesn't involve a new deployment, git tag, docker image tag, and it doesn't affect traces
- [x] Yes

## 4. Breaking change label

Does this PR have the `breaking-change` label?

- [x] This PR is not a breaking change
- [ ] Yes

### CODE REVIEW

**Comment by 0xmountaintop** [comment]
I could just review https://github.com/scroll-tech/go-ethereum/pull/982/commits/74a343077179ed2961c640951e3a972a807834f7 & https://github.com/scroll-tech/go-ethereum/pull/982/commits/c8c79c93742cc00a3095828e30fe383230a02acd?

(Assume https://github.com/scroll-tech/go-ethereum/pull/970 & https://github.com/scroll-tech/go-ethereum/pull/971 are ready)

**Comment by omerfirmak** [comment]
> I could just review [74a3430](https://github.com/scroll-tech/go-ethereum/commit/74a343077179ed2961c640951e3a972a807834f7) & [c8c79c9](https://github.com/scroll-tech/go-ethereum/commit/c8c79c93742cc00a3095828e30fe383230a02acd)?

Yes, exactly

File: miner/scroll_worker.go
Comment (0xmountaintop) [inline miner/scroll_worker.go]:
```suggestion
	// Only set the coinbase if our consensus engine is running (avoid spurious block rewards)
```

File: miner/scroll_worker.go
Comment (omerfirmak) [inline miner/scroll_worker.go:988]: this can block indefinitely if `mainLoop` is blocked on `asyncChecker.Check()`and cause a deadlock.

File: miner/scroll_worker.go
Comment (Thegaram) [inline miner/scroll_worker.go:501]: Why update `shouldCommit` here? Seems like it's always true

File: miner/scroll_worker.go
Comment (Thegaram) [inline miner/scroll_worker.go:517]: What happens if we're interrupted?

Even if that happens, we should detect that CCC was not run/stored on the last few blocks, and re-run after restart

File: miner/scroll_worker.go
Comment (omerfirmak) [reply miner/scroll_worker.go:501]: We might decide to let worker process txn pool as well after handling reorged txns.

File: rollup/ccc/logger.go
Comment (colinlyguo) [inline rollup/ccc/logger.go:228]: `err == nil` or intentional `err != nil`?

File: miner/scroll_worker.go
Comment (0xmountaintop) [inline miner/scroll_worker.go]: we can add a comment about what the returned value stands for

File: miner/scroll_worker.go
Comment (0xmountaintop) [inline miner/scroll_worker.go:725]: in production we should not invoke this?

File: miner/scroll_worker.go
Comment (0xmountaintop) [inline miner/scroll_worker.go:1010]: you mean because of the check in https://github.com/scroll-tech/go-ethereum/blob/378048c154c0f248ed9fe719dc9eb4e6f57f636d/miner/scroll_worker.go#L454?

**Comment by 0xmountaintop** [comment]
generally LGTM

File: miner/scroll_worker.go
Comment (colinlyguo) [inline miner/scroll_worker.go:61]: should we add reorg depth-related metrics?

File: miner/scroll_worker.go
Comment (colinlyguo) [inline miner/scroll_worker.go:1011]: Confused about cases in the force commit here. any examples?

File: rollup/ccc/libzkp/Cargo.toml
Comment (georgehao) [inline rollup/ccc/libzkp/Cargo.toml]: maybe 0.12.0?

**Review by colinlyguo — DISMISSED** [review]

File: eth/tracers/tracers.go
Comment (Thegaram) [inline eth/tracers/tracers.go]: Hmm why change this?

File: miner/scroll_worker.go
Comment (colinlyguo) [inline miner/scroll_worker.go:930]: What about adding an else branch to log the unexpected cases? e.g. L2 transactions not due to the reasons in the above conditional branches.

**Review by colinlyguo — DISMISSED** [review]

File: rollup/ccc/logger.go
Comment (Thegaram) [inline rollup/ccc/logger.go:80]: Will `currentEnv` be shared?

File: rollup/ccc/logger.go
Comment (Thegaram) [inline rollup/ccc/logger.go]: At what gas limit would this become unsafe?

File: rollup/ccc/logger.go
Comment (Thegaram) [inline rollup/ccc/logger.go:256]: Isn't it a duplicate with the call in `CaptureStateAfter` and `CaptureEnter`? Or are these called for different contexts?

File: rollup/ccc/async_checker.go
Comment (Thegaram) [inline rollup/ccc/async_checker.go:102]: When can this happen in practice?

File: miner/scroll_worker_test.go
Comment (Thegaram) [inline miner/scroll_worker_test.go:1029]: Are we just testing that at least 1 txs is included in the pending block?

File: miner/scroll_worker.go
Comment (Thegaram) [inline miner/scroll_worker.go:364]: Will this only return false if we had previously reorged this block?

File: miner/scroll_worker.go
Comment (Thegaram) [inline miner/scroll_worker.go:634]: Will this process a single block? Or all subsequent blocks too?

**Review by Thegaram — DISMISSED** [review]
LGTM, just a few questions

File: rollup/ccc/libzkp/Cargo.toml
Comment (colinlyguo) [reply rollup/ccc/libzkp/Cargo.toml]: The CCC version is in another version series. this pr will upgrade it from `v0.11.7` to `v0.11.8`.

File: miner/scroll_worker.go
Comment (colinlyguo) [reply miner/scroll_worker.go:61]: recorded by `blockReorgDropMeter`.

File: miner/scroll_worker.go
Comment (colinlyguo) [inline miner/scroll_worker.go]: Would `config.CCCMaxWorkers+1` be sufficient for this queue?

File: miner/scroll_worker.go
Comment (omerfirmak) [reply miner/scroll_worker.go:517]: > What happens if we're interrupted?
Nothing catastrophic, it is just a more graceful way of shutting down.

> Even if that happens, we should detect that CCC was not run/stored on the last few blocks, and re-run after restart

Yeap, this was added in https://github.com/scroll-tech/go-ethereum/pull/982/commits/057b2c10c3738b9662fd68f4c86c88c5e581a233

File: miner/scroll_worker.go
Comment (omerfirmak) [reply miner/scroll_worker.go:725]: It should be fine, `forceTestErr` is a noop in production.

File: miner/scroll_worker.go
Comment (omerfirmak) [reply miner/scroll_worker.go:1010]: Nope because of this check https://github.com/scroll-tech/go-ethereum/blob/378048c154c0f248ed9fe719dc9eb4e6f57f636d/miner/scroll_worker.go#L955-L959

File: rollup/ccc/logger.go
Comment (omerfirmak) [reply rollup/ccc/logger.go:80]: It will get overwritten when the snapshot is used for the first time, but it is safer to set it to nil here. Thanks! https://github.com/scroll-tech/go-ethereum/pull/982/commits/50636953bdcfc2bf0a7db4f648e85299e5242a2c

File: eth/tracers/tracers.go
Comment (omerfirmak) [reply eth/tracers/tracers.go]: to make sure `GetMemoryCopyPadded` wont error. The call site of `GetMemoryCopyPadded` is not really capable of handling errors.

File: rollup/ccc/logger.go
Comment (omerfirmak) [reply rollup/ccc/logger.go]: ~20M

File: rollup/ccc/logger.go
Comment (omerfirmak) [reply rollup/ccc/logger.go:256]: `CaptureEnter/CaptureStateAfter` are for CREATE/CREATE2 opcodes, `CaptureStart/CaptureEnd` are for implicit create txns with `To = nil`

File: rollup/ccc/async_checker.go
Comment (omerfirmak) [reply rollup/ccc/async_checker.go:102]: In sequencer it should never happen that is why I added the log to verify my assumption. In follower nodes it can happen during reorgs.

File: rollup/ccc/logger.go
Comment (Thegaram) [reply rollup/ccc/logger.go]: Got it. My only concern is that we'll increase gas limit in the future and forget about this. In that case, would we need to increase `memoryPadLimit` again?

File: miner/scroll_worker_test.go
Comment (omerfirmak) [reply miner/scroll_worker_test.go:1029]: Yeap, it is just there to make sure RPC nodes keep building their pending blocks.
This is not a new test tho, the diff looks weird because `TestSealBlockAfterCliquePeriod` was removed.

File: rollup/ccc/async_checker.go
Comment (Thegaram) [reply rollup/ccc/async_checker.go:102]: Oh, I was thinking this will happen during "graceful restart" (in `checkHeadRowConsumption`), is that not the case?

File: miner/scroll_worker.go
Comment (omerfirmak) [reply miner/scroll_worker.go:364]: Correct.

File: miner/scroll_worker.go
Comment (omerfirmak) [reply miner/scroll_worker.go:634]: This handles a single block only. `handleReorg` is responsible of reorging the entire faulty part of the chain.

File: miner/scroll_worker.go
Comment (omerfirmak) [reply miner/scroll_worker.go]: It might be, but if I am not mistaken it needs to be able to hold atleast 2 reorgs worth of events. It is a bit risky to lower this channel's buffer size because it might cause deadlocks.

File: rollup/ccc/async_checker.go
Comment (omerfirmak) [reply rollup/ccc/async_checker.go:102]: Yeah that would be a pain to deal with. Added error handling https://github.com/scroll-tech/go-ethereum/pull/982/commits/843f5fa12276c314b73ce5d424a57aae0641a56f

File: rollup/ccc/logger.go
Comment (omerfirmak) [reply rollup/ccc/logger.go]: Yeah that would be a pain to deal with. Added error handling https://github.com/scroll-tech/go-ethereum/pull/982/commits/843f5fa12276c314b73ce5d424a57aae0641a56f

File: rollup/ccc/async_checker.go
Comment (omerfirmak) [reply rollup/ccc/async_checker.go:102]: If I am not mistaken, it shouldn't happen in `checkHeadRowConsumption`. Because if `checkHeadRowConsumption` ends up spawning checker jobs, it will still do it for a continous chain of blocks. It should never skip a block in-between and spawn a task for the next one. (ie spawn for N, N+2 but skip N+1)

File: miner/scroll_worker.go
Comment (omerfirmak) [reply miner/scroll_worker.go:930]: I am worried about spamming the logs. But I can still add it if you think it will be useful

File: miner/scroll_worker.go
Comment (omerfirmak) [reply miner/scroll_worker.go]: The return value has no use. removed https://github.com/scroll-tech/go-ethereum/pull/982/commits/024a49c01c2d26efc279c8c4abbea7baff544640.

File: rollup/ccc/logger.go
Comment (Thegaram) [reply rollup/ccc/logger.go:256]: But we log access in both `CaptureEnter/CaptureStateAfter`, isn't that counting twice?

Comment (omerfirmak) [reply rollup/ccc/logger.go:256]: `CaptureEnter` counts the init code (ie the code that generates the contract code) while `CaptureStateAfter` counts the deployed bytecode (the actual contract code).

File: miner/scroll_worker.go
Comment (colinlyguo) [reply miner/scroll_worker.go:930]: I may misunderstand this. Will it be impossible to go into the `else` branch (i.e. only the expected case) as designed?

Comment (colinlyguo) [reply miner/scroll_worker.go:930]: I see. there are plenty of error types here. then no problem.

File: rollup/ccc/logger.go
Comment (Thegaram) [inline rollup/ccc/logger.go:197]: Now that you reduced `memoryPadLimit back to its original value, will we run into this error in practice? I guess it's not likely that this happens in `logPrecompileAccess`? But if it does, what impact will it have (since `logPrecompileAccess` just ignores the error)?

File: miner/scroll_worker.go
Comment (Thegaram) [reply miner/scroll_worker.go]: Why 2 reorgs?

Also, this is complex enough to warrant adding a comment here.

**Review by Thegaram — DISMISSED** [review]

File: miner/scroll_worker.go
Comment (colinlyguo) [reply miner/scroll_worker.go]: Why 2 reorgs? => same question. @omerfirmak

File: rollup/ccc/async_checker.go
Comment (colinlyguo) [inline rollup/ccc/async_checker.go:105]: any triggering example of the follower `nil` consumption block here? e.g. it receives some future blocks first or?

Comment (colinlyguo) [reply rollup/ccc/async_checker.go:105]: I may understand it now. if a future block comes, and then the previous block of this checker is reverted due to rust ccc failure, the `forkCtx` may be canceled. triggering early return of `checkerTask`.

File: miner/scroll_worker.go
Comment (omerfirmak) [reply miner/scroll_worker.go]: Oh, now that I take another look at it; I was wrong. Buffer size of 1 is enough. We ignore CCC errors from blocks that have an ancestor that fails CCC. A write operation to reorgCh will always find an empty channel. Either because we read the trigger from the channel and currently processing the reorg or there is no reorg happening at that point in time.

File: rollup/ccc/logger.go
Comment (omerfirmak) [reply rollup/ccc/logger.go:197]: > But if it does, what impact will it have

We might underestimate `copy` circuit usage and cause a reorg.

File: miner/scroll_worker.go
Comment (colinlyguo) [reply miner/scroll_worker.go]: what about adding some comments on the invariance guarantee here? it will help debug if the implementation breaks the assumption.

Comment (colinlyguo) [reply miner/scroll_worker.go]: btw, does the conclusion based on this commit: [make double-reorgs impossible](https://github.com/scroll-tech/go-ethereum/pull/982/commits/3ea2c59a736b33cbc3ec9a81444ac97ba7ed5349)?

Comment (colinlyguo) [reply miner/scroll_worker.go]: btw, I still cannot think of a case of the `reorgCh` deadlock here. Could you provide one?

**Review by colinlyguo — DISMISSED** [review]
LGTM. What about adding the bump version tag now? since the pr is almost ready to be merged.

File: miner/scroll_worker.go
Comment (omerfirmak) [reply miner/scroll_worker.go]: The main one was this: (which is no longer the case)

- Multiple blocks fail CCC and we don't ignore later ones like we do right now.
- First one gets written to the channel, second write operation blocks
- mainLoop is blocked while calling Check() so it can't read from the reorg channel
- For Check() to unblock a checker task needs to become available.
- For a checker task to become available, they need to be able to write to reorg channel 
- But they can't write to reorg channel, because mainLoop can't read from reorg channel to free up some space. Since it is blocked on Check()

**Review by colinlyguo — APPROVED** [review]

**Review by Thegaram — APPROVED** [review]

**Pull request merged**
